src/components/carbon-provider/top-modal-context.tsx

```diff
@@ -0,0 +1,49 @@
+import React, {
+  createContext,
+  useState,
+  useEffect,
+  useMemo,
+  ReactNode,
+} from "react";
+
+type TopModalContextProps = {
+  topModal: HTMLElement | null;
+};
+
+const TopModalContext = createContext<TopModalContextProps>({
+  topModal: null,
+});
+
+export const TopModalContextProvider = ({
+  children,
+}: {
+  children: ReactNode;
+}) => {
+  const [topModal, setTopModal] = useState<HTMLElement | null>(null);
+
+  // can't add the setter to the global list inside useEffect because that doesn't run until
+  // after the render. Use useMemo instead so that it only runs once.
+  useMemo(() => {
```

Comment:
https://reactjs.org/docs/hooks-reference.html#usememo

`You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.`

It looks like use of `useMemo` does not guarantee that it's going to run only once.

Reply:
Also I think it would actually be better to contain that logic inside of `modal-manager` or `useModalManager` files, just pass the the `setTopModal` method through the context

Reply:
thanks. I'd forgotten useMemo isn't a guarantee. I guess I'll once again have to rethink how to do this.

As for your other comment - just passing setTopModal through the context doesn't work, we need to keep a global record of all such setters in order to ensure that when a modal is opened or closed in one MFE, it updates the topModal context in all others. I'm open to moving the logic to ModalManager though if it makes sense - I'll look at that.

Reply:
You might be able to do something with `useRef` to create the equivalent of a `componentWillMount` hook.

```
useBeforeMount(setter) => {
  const mounted = useRef(false)

  if (!mounted.current) {
    if (!window.__CARBON_INTERNALS_MODAL_SETTER_LIST) {
      window.__CARBON_INTERNALS_MODAL_SETTER_LIST = [];
    }

    window.__CARBON_INTERNALS_MODAL_SETTER_LIST.push(setter);
  }
  mounted.current = true
}
```

Reply:
@robinzigmond I don't really understand what is the issue with passing down the `setTopModal` through the context to the `useModalManager` for the sake of having that logic contained in one place.